### PR TITLE
Refactor query parsing

### PR DESCRIPTION
**Pull request by shubhankarunhale**
Added dictionaries for queries in each function.

### CODE REVIEW

**Comment by ttt-77** [comment]
Can you change it to the format {'test_query_i' : {'query_str':, 'normalized_query_str':, 'correct_service':, 'correct_tables':}}?

**Comment by shubhankarunhale** [comment]
@ttt-77 Okay sure, updating the dictionary format. Also, please let me know if I should add a key with the num_query parameter as well in the inner dictionary.

**Comment by shubhankarunhale** [comment]
Hi @ttt-77 , I have updated the dictionary structure as per your request. Also added an additional key "num_query" to each query dictionary to enable iteration of the test queries. Can you please review?

File: tests/test_query_parsing.py [hunk 1]
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
...
-    # filter predicates connected by OR are in same set
-    correct_service = [{'colors02', 'objects00'}]
-    correct_tables = [{'colors02', 'objects00'}]
-    self._test_query(query_str1, config, normalized_query_str1, correct_fp, correct_service, correct_tables, 3)
-
-
+    
+    queries = {
+    "test_query_1":{
+        "query_str": '''
```
Comment (ttt-77) [inline tests/test_query_parsing.py:46]: Better to define constant parameters for these variables, like QUERY_STR = 'query_str'

File: tests/test_query_parsing.py [hunk 2]
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
...
-
-    # filter predicates connected by OR are in same set
-    correct_service = [{'colors02', 'objects00'}]
-    correct_tables = [{'colors02', 'objects00'}]
-    self._test_query(query_str1, config, normalized_query_str1, correct_fp, correct_service, correct_tables, 3)
-
-
+    
+    queries = {
+    "test_query_1":{
```
Comment (ttt-77) [inline tests/test_query_parsing.py:45]: start with 0

File: tests/test_query_parsing.py [hunk 3]
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
...
+                                "objects00.x_max, objects00.y_max) "
+                                "FROM colors02 JOIN objects00 ON colors02.frame = objects00.frame "
+                                "WHERE objects00.frame > 10000 OR objects00.y_max < 800"),
+        "correct_fp" : [['blobs_00.frame > 10000', 'objects00.y_max < 800']],
+        "correct_service" : [{'objects00'}],
+        "correct_tables" : [{'objects00'}],
+        "num_query":1
+    }
+}
+    for i in range(1,10):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:196]: use 'for i in range(len(queries)):'

File: tests/test_query_parsing.py [hunk 4]
```diff
@@ -40,174 +40,161 @@ async def test_nested_query(self):
...
+      self._test_query(queries["test_query_"+str(i)]["query_str"], config, queries["test_query_"+str(i)]["normalized_query_str"], queries["test_query_"+str(i)]["correct_fp"],queries["test_query_"+str(i)]["correct_service"], queries["test_query_"+str(i)]["correct_tables"], queries["test_query_"+str(i)]["num_query"])
```
Comment (ttt-77) [inline tests/test_query_parsing.py:197]: refactor '_test_query' to receive only two parameters "queries[f'test_query_{i}']" and config

File: tests/test_query_parsing.py [hunk 5]
```diff
@@ -252,44 +238,45 @@ async def test_correct_approximate_query(self):
     register_inference_services(aidb_engine, data_dir)
 
     config = aidb_engine._config
-
-    # approx aggregation
-    query_str1 = '''
+    
+    queries = {
+    # approx aggregation as a subquery
+    "test_query_1":{
```
Comment (ttt-77) [inline tests/test_query_parsing.py:244]: Same modification logic as above

**Comment by ttt-77** [comment]
Can you also help refactor 'test_udf_query' with the similar format? @shubhankarunhale

File: tests/test_query_parsing.py [hunk 6]
```diff
@@ -11,25 +11,33 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+QUERY_STR = "query_str"
```
Comment (ttt-77) [inline tests/test_query_parsing.py:14]: single quotation

File: tests/test_query_parsing.py [hunk 7]
```diff
@@ -40,174 +48,162 @@ async def test_nested_query(self):
...
+                                "FROM colors02 JOIN objects00 ON colors02.frame = objects00.frame "
+                                "WHERE objects00.frame > 10000 OR objects00.y_max < 800"),
+        CORRECT_FP : [['blobs_00.frame > 10000', 'objects00.y_max < 800']],
+        CORRECT_SERVICE : [{'objects00'}],
+        CORRECT_TABLES : [{'objects00'}],
+        NUM_QUERY:1
+    }
+}
+    for i in range(len(queries)):
+      self._test_query(queries["test_query_"+str(i)], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:206]: use "queries[f'test_query_{i}']"

File: tests/test_query_parsing.py [hunk 8]
```diff
@@ -252,44 +247,45 @@ async def test_correct_approximate_query(self):
...
-
+                             "CONFIDENCE 95%"),
+        CORRECT_FP :[["colors02.color IN (SELECT color FROM colors02 WHERE frame > 10000)"]],
+        CORRECT_SERVICE : [{'colors02'}],
+        CORRECT_TABLES : [{'colors02'}],
+        NUM_QUERY:2
+    }
+}
+    for i in range(len(queries)):
+      self._test_query(queries["test_query_"+str(i)], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:288]: same as above

File: tests/test_query_parsing.py [hunk 9]
```diff
@@ -11,25 +11,40 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+QUERY_STR = 'query_str'
```
Comment (ttt-77) [inline tests/test_query_parsing.py:14]: Add comments to separate different parts' parameters.Something like 'normal query extraction parameters', 'udf query extraction parameters'.

**Comment by shubhankarunhale** [comment]
Hi @ttt-77 , I have implemented all the review comments. Can you please review?

File: tests/test_query_parsing.py [hunk 10]
```diff
@@ -11,25 +11,43 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+#normal query extraction parameters
```
Comment (ttt-77) [inline tests/test_query_parsing.py:14]: space between '#' and the character

**Comment by ttt-77** [comment]
Others looks good to me.

**Comment by shubhankarunhale** [comment]
@ttt-77 , Okay great! I have added the space to the comments.

**Comment by ttt-77** [comment]
Have you merged the newest code from main branch? The test didn't pass.

**Comment by shubhankarunhale** [comment]
> Have you merged the newest code from main branch? The test didn't pass.

Yes, I have merged the latest code into my branch.  
The test is failing at 
`dataframe_sql, query_after_extraction = query.udf_query`. with a KeyError
I am rechecking my refactored code, and looking into the failure.

File: tests/test_query_parsing.py [hunk 11]
```diff
@@ -325,426 +330,427 @@ def _test_equality(test_query_list, config):
...
+        FILTER_PREDICATE: "(function__0 = TRUE) AND (function__1 = TRUE) AND (function__2 > function__3)"
       }
-    }
-    test_query_list.append(test_query)
-
-    test_query_list.append(test_query)
-
+    },
+    # test user defined function for exact aggregation query
+    "test_query_7" : {
```
Comment (ttt-77) [inline tests/test_query_parsing.py:542]: Remove this one

**Comment by shubhankarunhale** [comment]
@ttt-77 I have removed the the exact aggregation query

**Comment by shubhankarunhale** [comment]
@ttt-77  thanks for the input, the tests are passing now. Can you please take a look?

**Comment by ttt-77** [comment]
@ddkang Could you please review this PR?

File: tests/test_query_parsing.py [hunk 12]
```diff
@@ -11,25 +11,43 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+# normal query extraction parameters
+QUERY_STR = 'query_str'
+NORMALIZED_QUERY_STR = 'normalized_query_str'
+CORRECT_FP='correct_fp'
```
Comment (ddkang) [inline tests/test_query_parsing.py:17]: spaces after =

File: tests/test_query_parsing.py [hunk 13]
```diff
@@ -219,27 +225,26 @@ async def test_approximate_query_as_subquery(self):
     register_inference_services(aidb_engine, data_dir)
 
     config = aidb_engine._config
-
-    # approx aggregation as a subquery
-    query_str1 = '''
+    queries = {
+      # approx aggregation as a subquery
+      "query_str1":'''
```
Comment (ddkang) [inline tests/test_query_parsing.py:230]: spaces

File: tests/test_query_parsing.py [hunk 14]
```diff
@@ -11,25 +11,43 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+# normal query extraction parameters
+QUERY_STR = 'query_str'
+NORMALIZED_QUERY_STR = 'normalized_query_str'
+CORRECT_FP= 'correct_fp'
```
Comment (ttt-77) [inline tests/test_query_parsing.py:17]: space before =

**Comment by shubhankarunhale** [comment]
> @shubhankarunhale please tag the issue in development if this resolves an issue
@ddkang  I have added the reference to issue

File: tests/test_query_parsing.py [hunk 15]
```diff
@@ -252,60 +257,60 @@ async def test_correct_approximate_query(self):
...
-    # approx select as a subquery
-    query_str2 = '''
+                             "WHERE objects00.frame > (SELECT AVG(frame) FROM blobs_00) ERROR_TARGET 10% CONFIDENCE 95%"),
+ 
+        CORRECT_FP : [["blobs_00.frame > (SELECT AVG(frame) FROM blobs_00)"]],
+        
+        # filter predicates connected by OR are in same set
+        CORRECT_SERVICE: [{}],
+        CORRECT_TABLES: [{}],
+        NUM_QUERY: 2
```
Comment (ddkang) [inline tests/test_query_parsing.py:278]: Why is this hard coded?

Comment (ttt-77) [reply tests/test_query_parsing.py:278]: This is used to check the correctness of parsing results. The hard coded parts are desired results.

Comment (ddkang) [reply tests/test_query_parsing.py:278]: What does NUM_QUERY: 2 mean?

Comment (ttt-77) [reply tests/test_query_parsing.py:278]: The query contain a subquery, so there are two queries in total. Or the number of 'SELECT' in the query sentence.

Comment (ddkang) [reply tests/test_query_parsing.py:278]: This should probably be renamed to something like NUMBER_OF_SELECT_CLAUSES or something shorter

File: tests/test_query_parsing.py [hunk 16]
```diff
@@ -11,25 +11,43 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+# normal query extraction parameters
+QUERY_STR = 'query_str'
```
Comment (ddkang) [inline tests/test_query_parsing.py:15]: This entire thing should probably be an Enumerate

File: tests/test_query_parsing.py (hunk 15 above)
Comment (shubhankarunhale) [reply tests/test_query_parsing.py:278]: Okay sure, renaming it

File: tests/test_query_parsing.py (hunk 16 above)
Comment (shubhankarunhale) [reply tests/test_query_parsing.py:15]: Okay, sure.  Changing to an Enum

Comment (shubhankarunhale) [reply tests/test_query_parsing.py]: @ddkang Done with the suggested changes

File: tests/test_query_parsing.py [hunk 17]
```diff
@@ -8,28 +8,51 @@
...
+class QueryParameters(str, Enum):
+  # normal query extraction parameters
+  QUERY_STR = 'query_str'
+  NORMALIZED_QUERY_STR = 'normalized_query_str'
+  CORRECT_FP = 'correct_fp'
+  CORRECT_SERVICE = 'correct_service'
+  CORRECT_TABLES = 'correct_tables'
+  NUM_OF_SELECT_CLAUSES = 'num_of_select_clauses'
+
+  # udf query extraction parameters
```
Comment (ttt-77) [inline tests/test_query_parsing.py:24]: Use another class for udf query

Comment (shubhankarunhale) [reply tests/test_query_parsing.py:28]: Done, added another class.

File: tests/test_query_parsing.py [hunk 18]
```diff
@@ -8,28 +8,55 @@
...
+  NORMALIZED_QUERY_STR = 'normalized_query_str'
+  CORRECT_FP = 'correct_fp'
+  CORRECT_SERVICE = 'correct_service'
+  CORRECT_TABLES = 'correct_tables'
+  NUM_OF_SELECT_CLAUSES = 'num_of_select_clauses'
+  
+  def __str__(self) -> str:
+        return self.value
+
+class UDFQueryParameters(str, Enum):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:27]: UdfQueryParameters

File: tests/test_query_parsing.py [hunk 19]
```diff
@@ -40,174 +67,162 @@ async def test_nested_query(self):
...
-    correct_service = [{'colors02', 'objects00'}]
-    correct_tables = [{'colors02', 'objects00'}]
-    self._test_query(query_str1, config, normalized_query_str1, correct_fp, correct_service, correct_tables, 3)
-
-
+    
+       
+    queries = {
+    "test_query_0": {
+        QueryParameters.QUERY_STR: '''
```
Comment (ttt-77) [inline tests/test_query_parsing.py:74]: Can you reformat here and below? Start with a new line with four space like
QueryParameters.QUERY_STR:
        '''
        SELECT
        FROM
        '''

Comment (shubhankarunhale) [reply tests/test_query_parsing.py:74]: @ttt-77 Can you please let me know if I should reformat all the other fields (CORRECT_FP,etc) in this format as well, or just the query and normalized query fields?

Comment (ttt-77) [reply tests/test_query_parsing.py:74]: All

Comment (shubhankarunhale) [reply tests/test_query_parsing.py]: Done, reformatted all the dictionaries.

**Comment by ttt-77** [comment]
Looks good to me. Could you please review this PR? @ddkang

File: tests/test_query_parsing.py [hunk 20]
```diff
@@ -8,28 +8,55 @@
 from tests.inference_service_utils.inference_service_setup import \
     register_inference_services
 from tests.utils import setup_gt_and_aidb_engine
+from enum import Enum
```
Comment (ddkang) [inline tests/test_query_parsing.py:11]: isort

File: tests/test_query_parsing.py [hunk 21]
```diff
@@ -40,174 +67,220 @@ async def test_nested_query(self):
...
-    correct_service = [{'colors02', 'objects00'}]
-    correct_tables = [{'colors02', 'objects00'}]
-    self._test_query(query_str1, config, normalized_query_str1, correct_fp, correct_service, correct_tables, 3)
-
-
+    
+       
+    queries = {
+    "test_query_0": {
+        QueryParameters.QUERY_STR:
```
Comment (ddkang) [inline tests/test_query_parsing.py:74]: Why aren't these dataclasses instead of dicts actually?

Comment (shubhankarunhale) [reply tests/test_query_parsing.py:74]: I see no issues with moving to dataclasses, working on the changes.

**Comment by shubhankarunhale** [comment]
Done with the following changes:

- Replaced dictionaries with dataclasses in queries
- Removed constant enums as the query parameter constants are no longer redundant
- Refactored _test_query and test_udf_query methods to work with dataclasses
- Reformatted all the dataclass queries as per the previously mentioned format

File: tests/test_query_parsing.py [hunk 22]
```diff
@@ -11,25 +12,57 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
+# normal query dataclass
+@dataclass
+class TestQuery():
+    query_str: str
```
Comment (ttt-77) [inline tests/test_query_parsing.py:18]: two spaces

File: tests/test_query_parsing.py [hunk 23]
```diff
@@ -11,25 +12,57 @@
...
+    normalized_query_str: str
+    correct_fp: list
+    correct_service: list
+    correct_tables: list
+    num_of_select_clauses: int
+
```
Comment (ttt-77) [inline tests/test_query_parsing.py:24]: two blank lines

File: tests/test_query_parsing.py [hunk 24]
```diff
@@ -40,174 +73,164 @@ async def test_nested_query(self):
...
+        SELECT frame, object_id 
+        FROM colors02 AS cl
+        WHERE cl.object_id > (SELECT AVG(object_id) FROM objects00
+        WHERE frame > (SELECT AVG(frame) FROM blobs_00 WHERE frame > 500))
+        ''',
+        ("SELECT colors02.frame, colors02.object_id FROM colors02 "
+         "WHERE colors02.object_id > (SELECT AVG(object_id) FROM objects00 WHERE frame > "
+         "(SELECT AVG(frame) FROM blobs_00 WHERE frame > 500))"),
+        # test replacing column with root column in filtering predicate,
+        # the root column of 'colors02.object_id' is 'objects00'
```
Comment (ttt-77) [inline tests/test_query_parsing.py:131]: remove

File: tests/test_query_parsing.py [hunk 25]
```diff
@@ -40,174 +73,164 @@ async def test_nested_query(self):
...
+        [["objects00.object_id > (SELECT AVG(object_id) FROM objects00 " 
+          "WHERE frame > (SELECT AVG(frame) FROM blobs_00 WHERE frame > 500))"]],
+        # filter predicates connected by AND are in different set
```
Comment (ttt-77) [inline tests/test_query_parsing.py:134]: remove

File: tests/test_query_parsing.py [hunk 26]
```diff
@@ -219,27 +242,30 @@ async def test_approximate_query_as_subquery(self):
...
-    # approx select as a subquery
-    query_str2 = '''
-                 SELECT x_min FROM  objects00
-                 WHERE frame IN (SELECT frame FROM colors02 where color LIKE 'blue'
-                    RECALL_TARGET {RECALL_TARGET}% CONFIDENCE 95%;)
-                 AND table2.object_id > (SELECT AVG(ob.object_id) FROM objects00 AS ob WHERE frame > 500);
-                 '''
-
-    parsed_query = Query(query_str2, config)
+    parsed_query = Query(queries["query_str2"], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:268]: merge it with Query(queries["query_str1"], config)

File: tests/test_query_parsing.py [hunk 27]
```diff
@@ -11,25 +12,57 @@
...
+    num_of_select_clauses: int
+
 class QueryParsingTests(IsolatedAsyncioTestCase):
   def are_lists_equal(self, list1, list2):
     for sub_list1, sub_list2 in zip(list1, list2):
       assert Counter(sub_list1) == Counter(sub_list2)
 
-  def _test_query(self, query_str, config, normalized_query, correct_fp, correct_service, correct_tables, num_query):
-    query = Query(query_str, config)
+  def _test_query(self, test_query, config):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:52]: can you move this function and 'are_lists_equal' into dataclass?

File: tests/test_query_parsing.py [hunk 28]
```diff
@@ -252,60 +278,61 @@ async def test_correct_approximate_query(self):
...
-        assert len(dataframe_sql['udf_mapping']) == len(test_query['dataframe_sql']['udf_mapping'])
-        assert all(any(e1 == e2 for e2 in dataframe_sql['udf_mapping'])
-                  for e1 in test_query['dataframe_sql']['udf_mapping'])
-        assert dataframe_sql['select_col'] == test_query['dataframe_sql']['select_col']
-        filter_predicate = query.convert_and_connected_fp_to_exp(dataframe_sql['filter_predicate'])
-        if filter_predicate:
-          filter_predicate = filter_predicate.sql()
-
-        assert filter_predicate == test_query['dataframe_sql']['filter_predicate']
+    def _test_equality(test_query, config):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:322]: can you move this function into dataclass?

**Comment by shubhankarunhale** [comment]
@ttt-77 Done with the suggested changes, can you please review?

File: tests/test_query_parsing.py [hunk 29]
```diff
@@ -11,26 +12,79 @@
...
+    
   def are_lists_equal(self, list1, list2):
     for sub_list1, sub_list2 in zip(list1, list2):
       assert Counter(sub_list1) == Counter(sub_list2)
-
-  def _test_query(self, query_str, config, normalized_query, correct_fp, correct_service, correct_tables, num_query):
-    query = Query(query_str, config)
+      
+      
+  def _test_query(self, test_query, config):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:31]: only receive the parameter 'config', modify 'test_query' to 'self' below

File: tests/test_query_parsing.py [hunk 30]
```diff
@@ -11,26 +12,79 @@
...
+class UdfTestQuery():
+  query_str: str
+  query_after_extraction: str
+  dataframe_sql: DataframeSql
+  correct_service: list
+  correct_tables: list
+  num_of_select_clauses: int
+  
+    
+  def _test_equality(self,test_query, config):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:72]: same as above

**Comment by shubhankarunhale** [comment]
Done with the updates

File: tests/test_query_parsing.py [hunk 31]
```diff
@@ -40,174 +94,160 @@ async def test_nested_query(self):
...
+         "FROM colors02 JOIN objects00 ON colors02.frame = objects00.frame "
+         "WHERE objects00.frame > 10000 OR objects00.y_max < 800"),
+        [['blobs_00.frame > 10000', 'objects00.y_max < 800']], 
+        [{'objects00'}],
+        [{'objects00'}],
+        1
+    )
+}
+    for i in range(len(queries)):
+      queries[f'test_query_{i}']._test_query(queries[f'test_query_{i}'], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:250]: only config is needed

File: tests/test_query_parsing.py [hunk 32]
```diff
@@ -219,30 +259,29 @@ async def test_approximate_query_as_subquery(self):
...
+      "query_str2": 
+          '''
+          SELECT x_min 
+          FROM  objects00
+          WHERE frame IN (SELECT frame FROM colors02 where color LIKE 'blue'
+          RECALL_TARGET {RECALL_TARGET}% CONFIDENCE 95%;)
+          AND table2.object_id > (SELECT AVG(ob.object_id) FROM objects00 AS ob WHERE frame > 500);
+          '''
+    }
+    parsed_query = Query(queries["query_str1"]+queries["query_str2"], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:281]: why you add these two string together?

File: tests/test_query_parsing.py (hunk 31 above)
Comment (shubhankarunhale) [reply tests/test_query_parsing.py:250]: Sorry missed that one, removing it

File: tests/test_query_parsing.py (hunk 32 above)
Comment (shubhankarunhale) [reply tests/test_query_parsing.py:281]: The intent was to test to merge them into one check as the assert is raised even if one of them does not contain approximate query as a subquery. Please let me know if we need a loop instead.

Comment (ttt-77) [reply tests/test_query_parsing.py:281]: use a loop

Comment (shubhankarunhale) [reply tests/test_query_parsing.py:281]: Done, used a loop.

File: tests/test_query_parsing.py [hunk 33]
```diff
@@ -278,9 +278,10 @@ async def test_approximate_query_as_subquery(self):
           AND table2.object_id > (SELECT AVG(ob.object_id) FROM objects00 AS ob WHERE frame > 500);
           '''
     }
-    parsed_query = Query(queries["query_str1"]+queries["query_str2"], config)
-    with self.assertRaises(Exception):
-      _ = parsed_query.all_queries_in_expressions
+    for i in range(len(queries)):
+        parsed_query = Query(queries[f'query_str{i}'], config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:282]: two spaces for in-block code

File: tests/test_query_parsing.py [hunk 34]
```diff
@@ -278,9 +278,10 @@ async def test_approximate_query_as_subquery(self):
...
+        with self.assertRaises(Exception):
+        _ = parsed_query.all_queries_in_expressions
```
Comment (ttt-77) [inline tests/test_query_parsing.py:284]: Indentation

Comment (shubhankarunhale) [reply tests/test_query_parsing.py:284]: Done

**Comment by ttt-77** [comment]
@shubhankarunhale Check the running error.

**Comment by shubhankarunhale** [comment]
> @shubhankarunhale Check the running error.

Looking into it

**Comment by shubhankarunhale** [comment]
Done with fixes

File: tests/test_query_parsing.py [hunk 35]
```diff
@@ -11,26 +12,76 @@
 
 DB_URL = "sqlite+aiosqlite://"
 
-class QueryParsingTests(IsolatedAsyncioTestCase):
+# normal query dataclass
+@dataclass
+class TestQuery(IsolatedAsyncioTestCase):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:17]: class TestQuery:

File: tests/test_query_parsing.py [hunk 36]
```diff
@@ -11,26 +12,76 @@
...
   def are_lists_equal(self, list1, list2):
     for sub_list1, sub_list2 in zip(list1, list2):
       assert Counter(sub_list1) == Counter(sub_list2)
-
-  def _test_query(self, query_str, config, normalized_query, correct_fp, correct_service, correct_tables, num_query):
-    query = Query(query_str, config)
+      
+      
+  def _test_query(self, config):
+    query = Query((self.query_str), config)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:32]: query = Query(self.query_str, config)

File: tests/test_query_parsing.py [hunk 37]
```diff
@@ -11,26 +12,76 @@
...
     # test the number of queries
-    assert len(query.all_queries_in_expressions) == num_query
-    self.assertEqual(query.query_after_normalizing.sql_str, normalized_query)
+    assert len(query.all_queries_in_expressions) == self.num_of_select_clauses
+    self.assertEqual(query.query_after_normalizing.sql_str, self.normalized_query_str)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:35]: assert query.query_after_normalizing.sql_str == self.normalized_query_str

File: tests/test_query_parsing.py [hunk 38]
```diff
@@ -11,26 +12,76 @@
...
       or_fp = []
       for or_connected in and_connected:
         or_fp.append(or_connected.sql())
       and_fp.append(or_fp)
-    self.are_lists_equal(and_fp, correct_fp)
-    self.are_lists_equal(query.inference_engines_required_for_filtering_predicates, correct_service)
-    self.are_lists_equal(query.tables_in_filtering_predicates, correct_tables)
-
+    self.are_lists_equal(and_fp, self.correct_fp)
+    self.are_lists_equal(query.inference_engines_required_for_filtering_predicates, test_query.correct_service)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:43]: self.are_lists_equal(query.inference_engines_required_for_filtering_predicates, self.correct_service)

File: tests/test_query_parsing.py [hunk 39]
```diff
@@ -11,26 +12,76 @@
...
+    self.are_lists_equal(query.tables_in_filtering_predicates, self.correct_tables)  
+
+
+# udf query dataclasses       
+@dataclass
+class UdfMapping():
```
Comment (ttt-77) [inline tests/test_query_parsing.py:49]: class UdfMapping:

File: tests/test_query_parsing.py [hunk 40]
```diff
@@ -11,26 +12,76 @@
...
+  col_names: list
+  function_name: str
+  result_col_name: list
+
+
+@dataclass
+class DataframeSql():
```
Comment (ttt-77) [inline tests/test_query_parsing.py:56]: class DataframeSql:

File: tests/test_query_parsing.py [hunk 41]
```diff
@@ -11,26 +12,76 @@
...
+  udf_mapping: list
+  select_col: list
+  filter_predicate: str
+
+
+@dataclass
+class UdfTestQuery(IsolatedAsyncioTestCase):
```
Comment (ttt-77) [inline tests/test_query_parsing.py:63]: class UdfTestQuery:

File: tests/test_query_parsing.py [hunk 42]
```diff
@@ -11,26 +12,76 @@
...
+  query_str: str
+  query_after_extraction: str
+  dataframe_sql: DataframeSql
+
+
+  def _test_equality(self, config):
+    query = Query(self.query_str, config)
+    dataframe_sql, query_after_extraction = query.udf_query
+    self.assertEqual(query_after_extraction.sql_str, self.query_after_extraction)
```
Comment (ttt-77) [inline tests/test_query_parsing.py:72]: assert query_after_extraction.sql_str == self.query_after_extraction

File: tests/test_query_parsing.py [hunk 43]
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
...
-        'filter_predicate': "(function__2 = TRUE) AND (function__3 = TRUE) AND (output1 > 1000) AND (output2 > 640000)"
-      }
-    }
-    test_query_list.append(test_query)
-
-
+        DataframeSql(
+           [UdfMapping(
+               ['col__0', 'col__1'],
+               'max_function'
```
Comment (ttt-77) [inline tests/test_query_parsing.py:613]: 'max_function',

**Comment by ttt-77** [comment]
@shubhankarunhale Before requesting a review, please run tests on your local computer to ensure successful execution.

**Comment by shubhankarunhale** [comment]
Hi @ttt-77 , okay sure. I am actually facing an issue with sqlite on my local env, I'll focus on resolving that before any additional commits.

**Comment by shubhankarunhale** [comment]
@ttt-77 Done with fixes, local run is passing as well. Can you please review?

**Comment by ttt-77** [comment]
Looks good to me. Could you please review this PR? @ddkang

File: tests/test_query_parsing.py [hunk 44]
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
 
     register_inference_services(aidb_engine, data_dir)
     config = aidb_engine._config
-
-    test_query_list = []
+
```
Comment (ddkang) [inline tests/test_query_parsing.py:348]: spaces

File: tests/test_query_parsing.py [hunk 45]
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
...
+            UdfMapping(
+                ['col__5', 'col__6'],
+                'is_equal', 
+                ['function__1'])
+            ],
+            ['col__0', 'col__1', 'col__2', 'col__3'],
+            '(function__0 = TRUE) AND (function__1 = TRUE)'
+        )
+    ),
+
```
Comment (ddkang) [inline tests/test_query_parsing.py:422]: spaces

File: tests/test_query_parsing.py [hunk 46]
```diff
@@ -323,428 +345,400 @@ def _test_equality(test_query_list, config):
...
+            UdfMapping( 
+                ['col__0', 'col__7'],
+                'multiply_function',
+                ['function__3'])
+            ],
+            ['col__0', 'col__1', 'col__2'], 
+            "(function__0 = TRUE) AND (function__1 = TRUE) AND (function__2 > function__3)"
+        )
+    ),
+
```
Comment (ddkang) [inline tests/test_query_parsing.py:554]: spaces

**Review by ddkang — APPROVED** [review]

**Pull request merged**